src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -77,17 +107,190 @@ public String getScript() {
 
     /**
      * Method for getting the type.
-     * @return The type of script -- dynamic, indexed, or file.
+     * 
+     * @return The type of script -- inline, indexed, or file.
      */
     public ScriptType getType() {
-        return type;
+        return type == null ? DEFAULT_TYPE : type;
+    }
+
+    /**
+     * Method for getting language.
+     * 
+     * @return The language of the script to be compiled/executed.
+     */
+    public String getLang() {
+        return lang;
     }
 
     /**
      * Method for getting the parameters.
+     * 
      * @return The map of parameters the script will be executed with.
      */
     public Map<String, Object> getParams() {
         return params;
     }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        script = in.readString();
+        if (in.readBoolean()) {
+            type = ScriptType.readFrom(in);
+        }
+        lang = in.readOptionalString();
+        if (in.readBoolean()) {
+            params = in.readMap();
+        }
+        doReadFrom(in);
+    }
+
+    protected void doReadFrom(StreamInput in) throws IOException {
+        // For sub-classes to Override
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(script);
+        boolean hasType = type != null;
+        out.writeBoolean(hasType);
+        if (hasType) {
+            ScriptType.writeTo(type, out);
+        }
+        out.writeOptionalString(lang);
+        boolean hasParams = params != null;
+        out.writeBoolean(hasParams);
+        if (hasParams) {
+            out.writeMap(params);
+        }
+        doWriteTo(out);
+    }
+
+    protected void doWriteTo(StreamOutput out) throws IOException {
```

Comment:
can't subclasses just override writeTo and call super before doing anything else?